src/io.rs

```diff
@@ -341,6 +341,106 @@ impl Write for &'_ mut [u8] {
     }
 }
 
+/// A "cursor" over a mutable byte buffer.
+///
+/// This type provides a `consume()` function, which can be called repeatedly
+/// to take portions of the buffer. An internal cursor will track the location
+/// of the buffer. This method is used to implement [`Write`] for `Cursor`.
+///
+/// This type is useful when you want to feed a scratch buffer into a function
+/// that performs I/O operations on a buffer, and then extract how much of the
+/// buffer was read or written. This is especialy useful when used in
+/// conjunction with [`ToWire`]:
+/// ```
+/// # use manticore::io::*;
+/// # use manticore::protocol::wire::*;
+/// # struct MyMessage;
+/// # impl ToWire for MyMessage {
+/// #     fn to_wire<W: Write>(&self, mut w: W) -> Result<(), ToWireError> {
+/// #         w.write_bytes(&[1, 2, 3, 4]);
+/// #         Ok(())
+/// #     }
+/// # }
+/// let msg = MyMessage;
+/// let mut buf = [0; 256];
+///
+/// let mut cursor = Cursor::new(&mut buf);
+/// msg.to_wire(&mut cursor);
+///
+/// let msg_bytes = cursor.take_consumed_bytes();
+/// assert_ne!(msg_bytes.len(), 0);
+/// ```
+///
+/// [`Write`]: trait.Write.html
+/// [`ToWire`]: ../protocol/wire/trait.ToWire.html
+pub struct Cursor<'a> {
+    buf: &'a mut [u8],
+    // Invariant: cursor <= buf.len().
+    cursor: usize,
+}
+
+impl<'a> Cursor<'a> {
+    /// Creates a new `Cursor` for the given buffer.
+    pub fn new(buf: &'a mut [u8]) -> Self {
+        Self { buf, cursor: 0 }
+    }
+
+    /// Consumes `n` bytes from the underlying buffer.
+    ///
+    /// If `n` bytes are unavailable, `None` is returned.
+    pub fn consume(&mut self, n: usize) -> Option<&mut [u8]> {
+        if self.buf.len() - self.cursor < n {
+            return None;
+        }
+        let end = self.cursor + n;
```

Comment:
How about:

```rust
let end = self.cursor + n;
if end > self.buf.len() {
  return None;
}
```
Just reads a little cleaner and uses the `end` value instead of constructing the constraint independently.

Reply:
This is fixed in #20.